### PR TITLE
Evaluate state (Kin) quantities exploiting CasADi code-generation

### PR DESCRIPTION
**Pull request by FabioBergonti**
This PR decreases the computational time for evaluating jacobian, ...

### CODE REVIEW

**Comment by FabioBergonti** [comment]
I compared 4s of a kinematic simulation of an 8x8 mesh W/O codegen.
Using codegen the computational time is lower.

image

If we increase the length of the simulation 10s the gap increase even more

image

This PR could be beneficial also for #2

**Review by gabrielenava — APPROVED** [review]
interesting! so the mex file is generated every time you run a new simulation and then deleted at the end? Do you think it may be convenient to keep also the generated mex files somewhere or not?

File: +mystica/+state/@StateKinMBody/StateKinMBody.m [hunk 1]
```diff
@@ -56,7 +59,20 @@ function clearProperties(obj)
...
+        function generateMEX(obj)
+            opts = struct('main', true,'mex', true);
+            C = casadi.CodeGenerator('mystica_stateKin.c',opts);
+            C.add(obj.csdFn.Jc);
+            C.add(obj.csdFn.intJcV);
+            C.add(obj.csdFn.rC_from_jointsAngVelPJ_2_jointsAngVel0);
+            C.add(obj.csdFn.rC_from_mBodyTwist0_2_jointsAngVelPJ);
+            C.add(obj.csdFn.get_mBodyVelQuat0_from_mBodyTwist0)
+            C.generate();
+            mex mystica_stateKin.c -largeArrayDims
```
Comment (Giulero) [inline +mystica/+state/@StateKinMBody/StateKinMBody.m:74]: For curiosity :)
Once the function `mystica_stateKin` is generated, does it live in the workspace and be called in every part of the code?

File: +mystica/runSimDynRel.m [hunk 2]
```diff
@@ -59,6 +59,7 @@
     %% Saving Workspace
 
     clear ans k kVec motorsCurrent mBodyPosQuat_0 tout dataLiveStatistics
+    mystica.utils.deleteGeneratedMEX
```
Comment (Giulero) [inline +mystica/runSimDynRel.m:62]: Does it also work without parenthesis?

**Review by Giulero — APPROVED** [review]
Left a couple of comments for curiosity ;)
It looks fine to me btw!

File: +mystica/runSimDynRel.m (hunk 2 above)
Comment (FabioBergonti) [reply +mystica/runSimDynRel.m:62]: In matlab yes :)

File: +mystica/+state/@StateKinMBody/StateKinMBody.m (hunk 1 above)
Comment (Giulero) [reply +mystica/+state/@StateKinMBody/StateKinMBody.m:74]: Also, is it computationally expensive to generate it? 
I might understand why you don't check if the function already exists. 
To check if the content of the function changed, you have to generate it again :D

**Review by Giulero — APPROVED** [review]

**Comment by FabioBergonti** [comment]
@gabrielenava 

> interesting! so the mex file is generated every time you run a new simulation and then deleted at the end?

Exactly!

> Do you think it may be convenient to keep also the generated mex files somewhere or not?

The generated MEX files are model dependent. If I want to keep them I have to add a check to verify that the model is not changed.
For the moment I decided to regenerate them every time also because it requires only ~10s

File: +mystica/+state/@StateKinMBody/StateKinMBody.m (hunk 1 above)
Comment (FabioBergonti) [reply +mystica/+state/@StateKinMBody/StateKinMBody.m:74]: I generate the MEX files in the current folder. So, until you do not change the current directory, they can be called everywhere

**Comment by FabioBergonti** [comment]
@Giulero @gabrielenava 

> Also, is it computationally expensive to generate it?

~10s but it depends on the dimension of the problem

> I might understand why you don't check if the function already exists.
To check if the content of the function changed, you have to generate it again :D

The time-demanding operation is the generation of the MEX files
Theoretically, I can compare the `mystica_stateKin.c` file and if they are different I can generate the MEX file
image

**Comment by FabioBergonti** [comment]
Thanks @gabrielenava and @Giulero 
Merging!

**Pull request merged**